Mage/src/main/java/mage/players/PlayerImpl.java

```diff
@@ -163,9 +163,12 @@ public abstract class PlayerImpl implements Player, Serializable {
 
     // indicates that the spell with the set sourceId can be cast with an alternate mana costs (can also be no mana costs)
     // support multiple cards with alternative mana cost
-    protected Set<UUID> castSourceIdWithAlternateMana = new HashSet<>();
-    protected Map<UUID, ManaCosts<ManaCost>> castSourceIdManaCosts = new HashMap<>();
-    protected Map<UUID, Costs<Cost>> castSourceIdCosts = new HashMap<>();
+    //
+    // A card may be able to cast multiple way with multiple methods.
+    // The specific MageIdentifier should be checked, before checking null as a fallback.
+    protected Map<UUID, Set<MageIdentifier>> castSourceIdWithAlternateMana = new HashMap<>();
+    protected Map<UUID, Map<MageIdentifier, ManaCosts<ManaCost>>> castSourceIdManaCosts = new HashMap<>();
+    protected Map<UUID, Map<MageIdentifier, Costs<Cost>>> castSourceIdCosts = new HashMap<>();
```

Comment:
I'm sure this could be refactored to be a `Map<UUID, Map<MageIdentifier, AlternateCastingCost>>`, with `AlternateCastingCost` a new object (with a proper copy function). Did not feel necessary here.